File: tests/test_work_graph_store_race_safety.py
from __future__ import annotations

import os
import socket
import time
from pathlib import Path

import pytest

from entroly.work_graph_store import WorkGraphStateError, WorkGraphStore


def _store(tmp_path: Path) -> WorkGraphStore:
    return WorkGraphStore(
        "repo:race-safety",
        root=tmp_path / "state",
        lock_timeout_seconds=0.01,
        stale_lock_seconds=1.0,
    )


def _age(path: Path, seconds: float = 5.0) -> None:
    old = time.time() - seconds
    os.utime(path, (old, old))


def test_live_local_owner_is_not_reclaimed_only_because_lock_is_old(tmp_path: Path) -> None:
    store = _store(tmp_path)
    token = f"{socket.gethostname()}:{os.getpid()}:live-owner"
    store.lock_path.write_text(f"{token}\n0\n", encoding="utf-8")
    _age(store.lock_path)

    assert store._break_stale_lock() is False
    assert store.lock_path.exists()
    assert store._lock_token() == token


def test_dead_or_unparseable_old_owner_is_still_reclaimable(tmp_path: Path) -> None:
    store = _store(tmp_path)
    store.lock_path.write_text("dead-owner\n0\n", encoding="utf-8")
    _age(store.lock_path)

    assert store._break_stale_lock() is True
    assert not store.lock_path.exists()


def test_oversized_lock_metadata_fails_closed(tmp_path: Path) -> None:
    store = _store(tmp_path)
    store.lock_path.write_bytes(b"x" * 5000)

    with pytest.raises(WorkGraphStateError, match="lock"):
        store._lock_token()


def test_symlinked_lock_is_never_followed(tmp_path: Path) -> None:
    if os.name == "nt" or not hasattr(os, "symlink"):
        pytest.skip("symlink semantics differ on this platform")
    store = _store(tmp_path)
    outside = tmp_path / "outside-lock"
    outside.write_text("outside\n", encoding="utf-8")
    store.lock_path.symlink_to(outside)

    with pytest.raises(WorkGraphStateError, match="unsafe Work Graph lock"):
        store._lock_token()
    assert outside.read_text(encoding="utf-8") == "outside\n"


def test_symlinked_state_is_rejected_before_native_graph_loading(tmp_path: Path) -> None:
    if os.name == "nt" or not hasattr(os, "symlink"):
        pytest.skip("symlink semantics differ on this platform")
    store = _store(tmp_path)
    outside = tmp_path / "outside-state"
    outside.write_text("not-a-work-graph\n", encoding="utf-8")
    store.state_path.symlink_to(outside)

    with pytest.raises(WorkGraphStateError, match="unsafe Work Graph state"):
        store._load_unlocked()
    assert outside.read_text(encoding="utf-8") == "not-a-work-graph\n"

File: entroly-wasm/test_work_graph_store_race_safety.js
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');
const {
  WorkGraphStateError,
  WorkGraphStore,
} = require('./js/work_graph_store');

function assert(condition, message) {
  if (!condition) throw new Error(message || 'assertion failed');
}

function age(filePath, milliseconds = 5000) {
  const old = new Date(Date.now() - milliseconds);
  fs.utimesSync(filePath, old, old);
}

const root = fs.mkdtempSync(path.join(os.tmpdir(), 'entroly-work-store-race-'));
try {
  const store = new WorkGraphStore('repo:race-safety', {
    root: path.join(root, 'state'),
    lockTimeoutMs: 10,
    staleLockMs: 1000,
  });

  const liveToken = `${os.hostname()}:${process.pid}:live-owner`;
  fs.writeFileSync(store.lockPath, `${liveToken}\n0\n`, { mode: 0o600 });
  age(store.lockPath);
  assert(store.breakStaleLock() === false, 'old live-owner lock was reclaimed');
  assert(fs.existsSync(store.lockPath), 'live-owner lock disappeared');
  assert(store.readLockToken() === liveToken, 'live-owner token changed');
  fs.unlinkSync(store.lockPath);

  const foreignToken = `definitely-other-host:${process.pid}:foreign-owner`;
  fs.writeFileSync(store.lockPath, `${foreignToken}\n0\n`, { mode: 0o600 });
  age(store.lockPath);
  assert(store.breakStaleLock() === false, 'old foreign-host lock was reclaimed without proof');
  assert(fs.existsSync(store.lockPath), 'foreign-host lock disappeared');
  assert(store.readLockToken() === foreignToken, 'foreign-host token changed');
  fs.unlinkSync(store.lockPath);

  fs.writeFileSync(store.lockPath, 'dead-owner\n0\n', { mode: 0o600 });
  age(store.lockPath);
  assert(store.breakStaleLock() === true, 'dead old lock was not reclaimed');
  assert(!fs.existsSync(store.lockPath), 'dead old lock still exists');

  fs.writeFileSync(store.lockPath, 'x'.repeat(5000), { mode: 0o600 });
  let oversizedRejected = false;
  try { store.readLockToken(); }
  catch (error) { oversizedRejected = error instanceof WorkGraphStateError; }
  assert(oversizedRejected, 'oversized lock metadata was accepted');
  fs.unlinkSync(store.lockPath);

  if (process.platform !== 'win32') {
    const outsideLock = path.join(root, 'outside-lock');
    fs.writeFileSync(outsideLock, 'outside\n');
    fs.symlinkSync(outsideLock, store.lockPath);
    let symlinkLockRejected = false;
    try { store.readLockToken(); }
    catch (error) { symlinkLockRejected = error instanceof WorkGraphStateError; }
    assert(symlinkLockRejected, 'symlink lock was followed');
    assert(fs.readFileSync(outsideLock, 'utf8') === 'outside\n', 'outside lock target changed');
    fs.unlinkSync(store.lockPath);

    const outsideState = path.join(root, 'outside-state');
    fs.writeFileSync(outsideState, 'not-a-work-graph\n');
    fs.symlinkSync(outsideState, store.statePath);
    let symlinkStateRejected = false;
    try { store.loadUnlocked(); }
    catch (error) { symlinkStateRejected = error instanceof WorkGraphStateError; }
    assert(symlinkStateRejected, 'symlink state was followed');
    assert(fs.readFileSync(outsideState, 'utf8') === 'not-a-work-graph\n', 'outside state target changed');
  }

  console.log('Work Graph npm store race safety contract: PASS');
} finally {
  fs.rmSync(root, { recursive: true, force: true });
}
